### PR TITLE
Separate execution of proof steps from modification of proof state

### PR DESCRIPTION
**Pull request by nwatson22**
- `step_proof` no longer modifies proof, only produces list of `ProofResult`
- `advance_proof` now retrieves `ProofResult`s from `step_proof` and applies them using `Proof.commit()`.
- `ImpliesProof` and `APRProof` reworked to follow this scheme
- `APRProofResult`, `APRProofExtendResult`, `APRProofSubsumeResult`, `APRProofBoundedResult` for different outcomes of stepping an `APRProof`

### CODE REVIEW

**Review by tothtamas28 — COMMENTED** [review]
Please split this into two PR: one that refactors KCFG modifications, and another that refactors the provers after that.

**Comment by nwatson22** [comment]
> Please split this into two PR: one that refactors KCFG modification, and another one that refactors the provers.

@tothtamas28 The refactor to KCFG is here: https://github.com/runtimeverification/pyk/pull/909

**Comment by nwatson22** [comment]
The refactor to the KCFG is now split off from this PR and merged, and this PR now only deals with the refactor to Proof/Prover.

File: src/pyk/proof/reachability.py [hunk 1]
```diff
@@ -105,6 +126,20 @@ def __init__(
...
 
+    def commit(self, result: StepResult) -> None:
+        if isinstance(result, APRProofExtendResult):
+            self.kcfg.extend(result.extend_result, self.kcfg.node(result.node_id), logs=self.logs)
+        elif isinstance(result, APRProofSubsumeResult):
+            if result.csubst is None:
+                self._terminal.add(result.node_id)
+            else:
+                self.kcfg.create_cover(result.node_id, self.target, csubst=result.csubst)
+                _LOGGER.info(f'Subsumed into target node {self.id}: {shorten_hashes((result.node_id, self.target))}')
```
Comment (ehildenb) [inline src/pyk/proof/reachability.py:137]: I don't quite understand this logic. Should it be two different result types, one for `APRProofTerminalResult` and one for `APRProofSubsumeResult`?

File: src/pyk/proof/reachability.py [hunk 2]
```diff
@@ -639,94 +674,66 @@ def _may_subsume(self, node: KCFG.Node) -> bool:
...
-        self.proof.kcfg.extend(extend_result=extend_result, node=node, logs=self.proof.logs)
+        is_terminal = self.kcfg_explore.kcfg_semantics.is_terminal(curr_node.cterm)
+        target_is_terminal = (self.proof.target in self.proof._terminal,)
 
-    def step_proof(self) -> None:
-        self._check_all_terminals()
+        if target_is_terminal and self.always_check_subsumption:
+            csubst = self._check_subsume(curr_node)
+            if csubst is not None or is_terminal:
+                return [APRProofSubsumeResult(csubst=csubst, node_id=curr_node.id)]
```
Comment (ehildenb) [inline src/pyk/proof/reachability.py:723]: Here, we could break it into two cases, `APRProofSubsumeResult` and `APRProofTerminalResult`.

Comment (nwatson22) [reply src/pyk/proof/reachability.py]: Fixed.

File: src/pyk/proof/implies.py [hunk 3]
```diff
@@ -55,6 +55,16 @@ def __init__(
...
 
+    def commit(self, result: StepResult) -> None:
+        proof_type = type(self).__name__
+        if isinstance(result, ImpliesProofResult):
+            self.csubst = result.csubst
+            self.simplified_antecedent = result.simplified_antecedent
+            self.simplified_consequent = result.simplified_consequent
+            _LOGGER.info(f'{proof_type} finished {self.id}: {self.status}')
+        else:
+            raise ValueError('Incorrect result type')
```
Comment (ehildenb) [inline src/pyk/proof/implies.py:66]:
```suggestion
            raise ValueError(f'Incorrect result type, expected ImpliesProofResult: {result}')
```

File: src/pyk/proof/reachability.py [hunk 4]
```diff
@@ -105,6 +131,19 @@ def __init__(
...
+            self.kcfg.extend(result.extend_result, self.kcfg.node(result.node_id), logs=self.logs)
+        elif isinstance(result, APRProofSubsumeResult):
+            self.kcfg.create_cover(result.node_id, self.target, csubst=result.csubst)
+            _LOGGER.info(f'Subsumed into target node {self.id}: {shorten_hashes((result.node_id, self.target))}')
+        elif isinstance(result, APRProofTerminalResult):
+            self._terminal.add(result.node_id)
+        elif isinstance(result, APRProofBoundedResult):
+            self.add_bounded(result.node_id)
+        else:
+            raise ValueError('Incorrect result type')
```
Comment (ehildenb) [inline src/pyk/proof/reachability.py:145]:
```suggestion
            raise ValueError(f'Incorrect result type, expected APRProofResult: {result}')
```

**Review by ehildenb — APPROVED** [review]
Approved with some small suggestions. It would be good to test this with Kontrol downstream before merge. Also, please don't merge until https://github.com/runtimeverification/evm-semantics/pull/2302 goes through, as we have enough changes going on downstream in that PR already.

**Comment by PetarMax** [comment]
Previously, `always_check_subsumption` was used exclusively in the case the `target` node was not terminal (this is possible for claims about internal functions, for example). Now, it seems to be used always, and there doesn't appear to be a subsumption check if the `target` node is not terminal. I'm not sure that the original functionality is preserved.

Separately, we are seeing failures [in Kontrol](https://github.com/runtimeverification/kontrol/pull/402) after this pyk update that may or may not be related, and there seems to be a proof for which no subsumption check is being made.

@nwatson22 @ehildenb

**Pull request merged**